internal/training/grpo_engine/kernels.c

```
@@ -0,0 +1,764 @@
+    double sum = 0.0;
+    for (int i = 0; i < n; i++) { x[i] = expf(x[i] - max_val); sum += (double)x[i]; }
+    float inv = (float)(1.0 / sum);
+    for (int i = 0; i < n; i++) x[i] *= inv;
+}
+
+/* ─── Top-p Sampling ─── */
+int grpo_top_p_sample(const float *logits, int vocab_size, float temp, float top_p,
+                      unsigned int *rng_state) {
+    /* Temperature scaling + softmax */
+    float *probs = (float *)malloc(vocab_size * sizeof(float));
+    for (int i = 0; i < vocab_size; i++) probs[i] = logits[i] / temp;
```

Cppcheck (2.21.0)

[warning] If memory allocation fails, then there is a possible null pointer dereference

(nullPointerOutOfMemory)